--- a/mdns.h	2026-08-28 15:59:14.898624754 +0900
+++ b/mdns.h	2026-08-28 15:59:36.016494009 +0900
@@ -1501,7 +1501,7 @@
 mdns_record_parse_ptr(const void* buffer, size_t size, size_t offset, size_t length,
                       char* strbuffer, size_t capacity) {
 	// PTR record is just a string
-	if ((size >= offset + length) && (length >= 2))
+	if ((offset <= size && length <= size - offset) && (length >= 2))
 		return mdns_string_extract(buffer, size, &offset, strbuffer, capacity);
 	mdns_string_t empty = {0, 0};
 	return empty;
@@ -1518,7 +1518,7 @@
 	// 2 bytes network-order unsigned weight
 	// 2 bytes network-order unsigned port
 	// string: discovery (domain) name, minimum 2 bytes when compressed
-	if ((size >= offset + length) && (length >= 8)) {
+	if ((offset <= size && length <= size - offset) && (length >= 8)) {
 		const uint16_t* recorddata = (const uint16_t*)MDNS_POINTER_OFFSET_CONST(buffer, offset);
 		srv.priority = mdns_ntohs(recorddata++);
 		srv.weight = mdns_ntohs(recorddata++);
@@ -1537,7 +1537,7 @@
 #ifdef __APPLE__
 	addr->sin_len = sizeof(struct sockaddr_in);
 #endif
-	if ((size >= offset + length) && (length == 4))
+	if ((offset <= size && length <= size - offset) && (length == 4))
 		memcpy(&addr->sin_addr.s_addr, MDNS_POINTER_OFFSET(buffer, offset), 4);
 	return addr;
 }
@@ -1550,7 +1550,7 @@
 #ifdef __APPLE__
 	addr->sin6_len = sizeof(struct sockaddr_in6);
 #endif
-	if ((size >= offset + length) && (length == 16))
+	if ((offset <= size && length <= size - offset) && (length == 16))
 		memcpy(&addr->sin6_addr, MDNS_POINTER_OFFSET(buffer, offset), 16);
 	return addr;
 }
@@ -1560,10 +1560,7 @@
                       mdns_record_txt_t* records, size_t capacity) {
 	size_t parsed = 0;
 	const char* strdata;
-	size_t end = offset + length;
-
-	if (size < end)
-		end = size;
+	size_t end = (offset <= size && length <= size - offset) ? offset + length : size;
 
 	while ((offset < end) && (parsed < capacity)) {
 		strdata = (const char*)MDNS_POINTER_OFFSET(buffer, offset);
